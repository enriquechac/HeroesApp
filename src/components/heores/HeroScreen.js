import React, { useMemo } from 'react';
import { Redirect, useParams } from 'react-router';
import { heroImages } from '../../helpers/heroImages';
import { getHeroesById } from '../../selectors/getHeroesById';

// import batman from '../../assets/heroes/dc-batman.jpg';

// const heroImages = require.context('../../assets/heroes', true);

export const HeroScreen = ({ history }) => {
    const { heroeId } = useParams();
    const hero = useMemo(() => getHeroesById(heroeId), [heroeId]);

    // console.log('Hero: ', hero, 'HeroId: ', heroeId);

    if (!hero) {
        return <Redirect to="/" />;
    }

    const {
        superhero,
        id,
        publisher,
        alter_ego,
        first_appearance,
        characters,
    } = hero;

    const handleReturn = () => {
        if (history.length <= 2) {
            history.push('/');
        } else {
            history.goBack();
        }
    };

    return (
        <div>
            <h1>HeroScreen</h1>
            <hr />
            <div className="row mt-5 animate__animated animate__fadeIn">
                <div className="col-4">
                    <img
                        // src={`../assets/heroes/${id}.jpg`} / Usando assets en la carpeta public
                        // src={batman} / Al usar import
                        src={heroImages(`./${id}.jpg`).default}
                        alt={superhero}
                        className="img-thumbnail"
                    />
                </div>
                <div className="col-8">
                    <h3>{superhero}</h3>
                    <ul className="list-goup list-group-flush">
                        <li className="list-group-item">
                            <b>Alter ego: {alter_ego}</b>
                        </li>
                        <li className="list-group-item">
                            <b>Publisher: {publisher}</b>
                        </li>
                        <li className="list-group-item">
                            <b>First Appearance: {first_appearance}</b>
                        </li>
                    </ul>
                    <h5>Characters</h5>
                    <p>{characters}</p>
                    <button
                        className="btn btn-outline-info"
                        onClick={handleReturn}
                    >
                        Return
                    </button>
                </div>
            </div>
        </div>
    );
};
