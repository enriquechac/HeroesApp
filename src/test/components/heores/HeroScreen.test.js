import { mount, shallow } from 'enzyme';
import { MemoryRouter, Redirect, Route } from 'react-router';
import { HeroScreen } from '../../../components/heores/HeroScreen';

describe('Pruebas en el componente <HeroScreen/>', () => {
    const historyMock = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn(),
    };

    test('Debe de mostrar el componente redirect si no hay argumentos en el URL', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={historyMock} />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Redirect').exists()).toBe(true);
    });

    test('Debe de mostrar un heroe si el parametro existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroeId" component={HeroScreen} />
            </MemoryRouter>
        );
        expect(wrapper.find('.row').exists()).toBe(true);
    });

    test('Debe de regresar a la pantalla anterior con push ', () => {
        const historyMock = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn(),
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route
                    path="/hero/:heroeId"
                    component={() => <HeroScreen history={historyMock} />}
                />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();

        expect(historyMock.push).toHaveBeenCalledWith('/');
        expect(historyMock.goBack).not.toHaveBeenCalled();
    });

    test('Debe de regresar a la pantalla anterior goBack ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route
                    path="/hero/:heroeId"
                    component={() => <HeroScreen history={historyMock} />}
                />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();

        expect(historyMock.push).toHaveBeenCalledTimes(0);
        expect(historyMock.goBack).toHaveBeenCalled();
    });

    test('Debe de llamar el redirect si el hero no existe', () => {
        const historyMock = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn(),
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/-------']}>
                <Route
                    path="/hero/:heroeId"
                    component={() => <HeroScreen history={historyMock} />}
                />
            </MemoryRouter>
        );
        expect(wrapper.text()).toBe('');
    });
});
