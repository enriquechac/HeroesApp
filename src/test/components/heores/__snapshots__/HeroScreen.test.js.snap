// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <HeroScreen/> Debe de mostrar el componente redirect si no hay argumentos en el URL 1`] = `""`;
